### PR TITLE
[WIP] Add TraceState.

### PR DESCRIPTION
**Pull request by g-easy**

### CODE REVIEW

**Comment by g-easy** [comment]
It's not clear to me what this should look like in C++.

In the Java API, it's abstract: [TraceState](https://github.com/open-telemetry/opentelemetry-java/blob/30e3bc8c6348d2c5fa857eb8b065ff891fd971f5/api/src/main/java/io/opentelemetry/trace/TraceState.java) and [SpanContext](https://github.com/open-telemetry/opentelemetry-java/blob/30e3bc8c6348d2c5fa857eb8b065ff891fd971f5/api/src/main/java/io/opentelemetry/trace/SpanContext.java)

File: api/include/opentelemetry/trace/trace_state.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    if (key.empty() || key.length() > kKeyMaxSize || !IsNumberOrDigit(key[0]))
+    {
+      return false;
+    }
+    int ats     = 0;
+    const int n = key.length();
+    for (int i = 0; i < n; ++i)
+    {
+      char c = key[i];
+      if (!IsNumberOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/')
```
Comment (reyang) [inline api/include/opentelemetry/trace/trace_state.h]: Vendor name has a size limitation of 13 characters according to the [spec](https://www.w3.org/TR/trace-context/#key).

File: api/include/opentelemetry/trace/trace_state.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // TODO: IsValidValue
+
+private:
+  static bool IsNumberOrDigit(char c) { return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9'); }
```
Comment (reyang) [inline api/include/opentelemetry/trace/trace_state.h]: IsLowerCaseAlphaOrDigit?

**Comment by reyang** [comment]
I checked the OT Python code:
1. TraceState implementation is in the API package [here](https://github.com/open-telemetry/opentelemetry-python/blob/c2ba0653518a3587c0cabf8b94fad374c9c53e56/opentelemetry-api/src/opentelemetry/trace/__init__.py#L279).
2. HTTP implementation (parser and serializer) of TraceContext is in the API package [here](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/trace/propagation/tracecontexthttptextformat.py#L145).

**Comment by Tianlin-Zhao** [comment]
I am currently working on context propagation and realized that I need methods from trace states to construct default trace states as well as inserting values for new trace states. Could you implement those as well?

File: api/include/opentelemetry/trace/span_context.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+// TraceFlags, TraceState, and whether it has a remote parent.
+class SpanContext final
+{
+public:
+  // An invalid SpanContext.
+  SpanContext() noexcept : trace_state_(new TraceState) {}
+
+  // TODO
+  //
+  // static SpanContext Create(TraceId traceId, SpanId spanId, TraceFlags traceFlags, TraceState
```
Comment (Tianlin-Zhao) [inline api/include/opentelemetry/trace/span_context.h:42]: Are these definitions going to be up eventually? I need these for my HttpTraceConotext implementation. Thanks

File: api/include/opentelemetry/trace/trace_state.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    virtual nostd::string_view value() = 0;
+  };
+
+  // An empty TraceState.
+  TraceState() noexcept = default;
+
+  virtual ~TraceState() = default;
+
+  // Returns false if no such key, otherwise returns true and populates value.
+  virtual bool Get(nostd::string_view key, nostd::string_view *value) const noexcept
```
Comment (Tianlin-Zhao) [inline api/include/opentelemetry/trace/trace_state.h:62]: Do we have set functions for Trace State any where? Could you have that functionality implemented? Thank you

**Comment by reyang** [comment]
Superseded by #234.

**Pull request closed without merging**